File: Word-Formed/src/com/google/corrigan/owen/wordformed/Dropbox.java
package com.google.corrigan.owen.wordformed;

import java.util.LinkedList;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.Typeface;
import android.util.Log;

public class Dropbox {
	private int border = 5;
	private RectF dragBorder, dragFill;
	private LinkedList<DraggableBox> tiles = new LinkedList<DraggableBox>();
	@SuppressWarnings("unused")
	private final String TAG = "DRAGGABLEBOX";

	Paint paint = new Paint();
	String instructions;

	public Dropbox(int x, int y, int width, int height) {
		dragBorder = new RectF(x, y, x + width, y + height);
		dragFill = new RectF(x + border, y + border, x + width - border, y + height - border);
		if(y == 325) {
		    instructions = "Save letters here";
		} else {
			instructions = "Form words here";
		}
		paint.setColor(R.color.black);
		paint.setTextSize(30);
		paint.setTypeface(Typeface.MONOSPACE);
		paint.setAlpha(150);
	}

	public String tilesToString() {
		String word = "";
		for(DraggableBox tile: tiles) {
			word += tile.getLetter();
		}
		Log.d("DropBox", word);
		return word;
	}

	public int getScore() {
		int val = 0;
		for(DraggableBox tile: tiles) {
			val += tile.getValue();
		}
		return val;
	}

	//Method for Adding tiles to the Dropbox
	public void add(DraggableBox d) {
		//size of 7 tiles held.
		if(tiles.size() == 7) {
			DraggableBox tmp = getFirst();
			tmp.move(-100, dragBorder.bottom - 75);
			tiles.remove(tmp);
		}
		tiles.add(d);
		Button.checkWord();
		updatePositions();
	}

	public void updatePositions() {
		for(int i = 0; i < tiles.size(); i++) {
			tiles.get(i).move(i*65+20, dragBorder.bottom - 75);
        }
	}

	public void remove(DraggableBox d) {
		tiles.remove(d);
		Button.checkWord();
		updatePositions();
	}

	@SuppressWarnings("unchecked")
	public synchronized void removeAll() {
		LinkedList<DraggableBox> clone = (LinkedList<DraggableBox>) tiles.clone();
		for(DraggableBox tile: clone) {
			DraggableBox tmp = getFirst();
			tmp.move(-100, dragBorder.bottom - 60);
			tiles.remove(tmp);		
		}
		Button.opaque();
		updatePositions();
	}

	public DraggableBox getFirst() {
		return tiles.get(0);
	}

	public void draw(Canvas canvas) {
		Paint dragRectangle = new Paint();
		dragRectangle.setColor(Color.BLACK);
		canvas.drawRect(dragBorder, dragRectangle);
		dragRectangle.setColor(Color.rgb(68, 89, 108));
		canvas.drawRect(dragFill, dragRectangle);
		canvas.drawText(instructions, dragBorder.left + 80, dragBorder.top + 60, paint);
	}

	public boolean contains(float x, float y) {
		return dragBorder.contains(x, y);
	}
}

File: app/src/main/java/com/google/corrigan/owen/wordformed/Dropbox.kt
package com.google.corrigan.owen.wordformed

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.Typeface
import android.util.Log
import java.util.LinkedList

class Dropbox(x: Int, y: Int, width: Int, height: Int) {
    private val border = 5
    private val dragBorder: RectF = RectF(x.toFloat(), y.toFloat(), (x + width).toFloat(), (y + height).toFloat())
    private val dragFill: RectF = RectF((x + border).toFloat(), (y + border).toFloat(), (x + width - border).toFloat(), (y + height - border).toFloat())
    private val tiles = LinkedList<DraggableBox>()
    private val paint = Paint()
    private val instructions: String

    var onTilesChanged: (() -> Unit)? = null

    init {
        instructions = if (y == 325) {
            "Save letters here"
        } else {
            "Form words here"
        }
        paint.color = Color.BLACK
        paint.textSize = 30f
        paint.typeface = Typeface.MONOSPACE
        paint.alpha = 150
    }

    fun tilesToString(): String {
        val word = StringBuilder()
        for (tile in tiles) {
            word.append(tile.getLetter())
        }
        val result = word.toString()
        Log.d("DropBox", result)
        return result
    }

    fun getScore(): Int {
        var valSum = 0
        for (tile in tiles) {
            valSum += tile.getValue()
        }
        return valSum
    }

    fun add(d: DraggableBox) {
        if (tiles.size == 7) {
            val tmp = getFirst()
            tmp.move(-100f, dragBorder.bottom - 75f)
            tiles.remove(tmp)
        }
        tiles.add(d)
        onTilesChanged?.invoke()
        updatePositions()
    }

    fun updatePositions() {
        for (i in tiles.indices) {
            tiles[i].move(i * 65f + 20f, dragBorder.bottom - 75f)
        }
    }

    fun remove(d: DraggableBox) {
        tiles.remove(d)
        onTilesChanged?.invoke()
        updatePositions()
    }

    @Synchronized
    fun removeAll() {
        val clone = LinkedList(tiles)
        for (tile in clone) {
            val tmp = getFirst()
            tmp.move(-100f, dragBorder.bottom - 60f)
            tiles.remove(tmp)
        }
        onTilesChanged?.invoke()
        updatePositions()
    }

    fun getFirst(): DraggableBox {
        return tiles.first
    }

    fun draw(canvas: Canvas) {
        val dragRectangle = Paint()
        dragRectangle.color = Color.BLACK
        canvas.drawRect(dragBorder, dragRectangle)
        dragRectangle.color = Color.rgb(68, 89, 108)
        canvas.drawRect(dragFill, dragRectangle)
        canvas.drawText(instructions, dragBorder.left + 80, dragBorder.top + 60, paint)
    }

    fun contains(x: Float, y: Float): Boolean {
        return dragBorder.contains(x, y)
    }
}
